File: src/main/scala/com/datastax/home_spotify/exercises/BaseExercise.scala
package com.datastax.home_spotify.exercises

import com.datastax.home_spotify.exercises.Constants._
import com.datastax.spark.connector.cql.CassandraConnector
import org.apache.spark.{SparkContext, SparkConf}

trait BaseExercise {

  def buildSparkContext(exerciseName: String):SparkContext = {

    val conf = new SparkConf(true)
      .setAppName(exerciseName)
      .setMaster(LOCAL_MODE)
      .set(CASSANDRA_HOST_NAME_PARAM, CASSANDRA_IP)


    if (Constants.TABLES.contains(exerciseName)) {
      CassandraConnector(conf).withSessionDo {
        session => session.execute(s"TRUNCATE ${Schema.KEYSPACE}.${Constants.TABLES(exerciseName)}")
      }
    }

    new SparkContext(conf)
  }
}

File: src/main/java/fr/ippon/home_spotify/exercises/BaseExercise.java
package fr.ippon.home_spotify.exercises;

import com.datastax.driver.core.Session;
import com.datastax.spark.connector.cql.CassandraConnector;
import com.datastax.spark.connector.cql.CassandraConnectorConf;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Unit;
import scala.runtime.AbstractFunction1;
import scala.tools.nsc.typechecker.Implicits;

import java.io.Serializable;

import static fr.ippon.home_spotify.exercises.Constants.*;
import static fr.ippon.home_spotify.exercises.Schema.*;

public class BaseExercise {

    protected static SparkConf buildScalaSparkConf(String exerciseName) {
        return new SparkConf()
                .setMaster(LOCAL_MODE)
                .setAppName(exerciseName)
                .set(CASSANDRA_HOST_NAME_PARAM, CASSANDRA_IP);
    }

    protected static JavaSparkContext buildSparkContext(String exerciseName) {
        SparkConf conf = buildScalaSparkConf(exerciseName);

        new CassandraConnector(CassandraConnectorConf.apply(conf)).withSessionDo(buildTruncateLambda(exerciseName));

        return new JavaSparkContext(conf);
    }

    protected static TruncateTables buildTruncateLambda(final String exerciseName) {
        return new TruncateTables() {

            @Override
            public Unit apply(Session session) {
                if (TABLES.containsKey(exerciseName)) {
                    session.execute("TRUNCATE " + KEYSPACE + "." + TABLES.get(exerciseName));
                }
                return null;
            }
        };
    }

    public static abstract class TruncateTables extends AbstractFunction1<Session, Unit> implements Serializable {

    }
}
